processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeDeriver.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.apm.processor.zipkin;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.hawkular.apm.api.model.Property;
+import org.hawkular.apm.api.model.events.CompletionTime;
+import org.hawkular.apm.server.api.model.zipkin.Annotation;
+import org.hawkular.apm.server.api.model.zipkin.Span;
+import org.hawkular.apm.server.api.services.SpanCache;
+import org.hawkular.apm.server.api.task.AbstractProcessor;
+import org.hawkular.apm.server.api.task.RetryAttemptException;
+import org.jboss.logging.Logger;
+
+/**
+ * Deriver for completion time of trace reported by zipkin instrumentation.
+ *
+ * @author Pavol Loffay
+ */
+@Singleton
+public class CompletionTimeDeriver extends AbstractProcessor<CompletionTimeProcessing, CompletionTime>{
+
+    private static final Logger log = Logger.getLogger(CompletionTimeDeriver.class.getName());
+
+    private SpanCache spanCache;
+
+
+    public CompletionTimeDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTime processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        Span rootSpan = completionTimeProcessing.getRootSpan();
+
+        List<Span> trace = spanCache.getTrace(tenantId, rootSpan.getTraceId());
+        Annotation lastAnnotation = extractLastAnnotation(trace);
+
+        /**
+         * TODO HWKAPM-348
+         *
+         * When there is new information recorded in trace (annotation with bigger timestamp)
+         * then wait for a fixed amount of time {@link #getRetryDelay(List, int)} to calculate
+         * trace completion time for all reported spans (async span can be recorded aby time after root
+         * span has been recorded). This approach does not deal with long run async spans,
+         * because it waits only for a fixed amount of time.
+         */
+        if (completionTimeProcessing.getLastTimestamp() == null ||
+                completionTimeProcessing.getLastTimestamp() < lastAnnotation.getTimestamp()) {
+            completionTimeProcessing.setLastTimestamp(lastAnnotation.getTimestamp());
+            return null;
```

Comment:
The way the event processing mechanism currently works is that each processor has one input type and one output type. So there is the initiator processor, which has Span has input and CompletionTimeProcessing as output, and then you will need another processor that converts the CompletionTimeProcessing ultimately to a CompletionTime event.

The problem is that the CompletionTimeProcessing may go through some iterations before it is ready to be converted to CompletionTime.

The issue is that at the moment, this processor is attempting to be the one that performs the interation, as well as the one that ultimately produces the CompletionTime event. The statement above modifies the CompletionTimeProcessing object, but does not then do anything with it.

You need to create another processor which just has responsibility for updating the CompletionTimeProcessing object and publishing it. The change this class so that it just checks for the processing object to determine if it should be converted to CompletionTime.

Not sure if I have explained this clearly, so ping me if you want to discuss.